### ISSUE TITLE
When gathering information Clang/windows errors reports ware not reported by build2 as errors

### ISSUE THREAD

**Issue by Klaim** [build2/build2#281]
At some point while working using clang on Windows (using the default msvc toolchain) I ended up with adding `-fPIC` to `config.cc.coptions` and hit the error below which was confusing because I didnt see the relationship with that flag and the output from `build2`. After realizing it was that flag, I setup a repro case:

```
bdep new somelib -t lib
cd somelib
bdep init -C ../build-clang cc config.cxx=clang++ config.c=clang config.cc.coptions+="-fPIC"
```
which lead to:
```
initializing in project E:\tests\build2-clangmsvc-fpic\somelib\
error: unable to extract MSVC information from clang@C:\Program Files\LLVM\bin\clang.exe
```

With more verbosity:
```
$ bdep init -C ../build-clang cc config.cxx=clang++ config.c=clang config.cc.coptions+="-fPIC" -V
mkdir C:\Users\Klaim\AppData\Local\Temp\bdep-2776-0\
initializing in project E:\tests\build2-clangmsvc-fpic\somelib\
mkdir E:\tests\build2-clangmsvc-fpic\somelib\.bdep\
E:\tools\build2\build2-install\bin\bpkg --verbose 3 create -d E:\tests\build2-clangmsvc-fpic\build-clang --type target --no-host-config --no-build2-config cc config.cxx=clang++ config.c=clang config.cc.coptions+=-fPIC
mkdir -p E:\tests\build2-clangmsvc-fpic\build-clang\
E:\tools\build2\build2-install\bin\b --verbose 3 config.cxx=clang++ config.c=clang config.cc.coptions+=-fPIC "create('E:\tests\build2-clangmsvc-fpic\build-clang\', cc)"
mkdir E:\tests\build2-clangmsvc-fpic\build-clang\build\
cat >E:\tests\build2-clangmsvc-fpic\build-clang\build\bootstrap.build
cat >E:\tests\build2-clangmsvc-fpic\build-clang\build\root.build
cat >E:\tests\build2-clangmsvc-fpic\build-clang\buildfile
clang -v
clang -fPIC -dumpmachine
clang -fPIC -x c -v -E -
error: unable to extract MSVC information from clang@C:\Program Files\LLVM\bin\clang.exe
rmdir -r C:\Users\Klaim\AppData\Local\Temp\bdep-2776-0\
```

I see that `-fPIC` was added to the commands used to retrieve the compiler information. I dont know if it's a bug or not(?) but I assumed it was on purpose for some configuration-specific handling reason. 
The issue therefore comes from that command:
```
$ clang -fPIC -x c -v -E -
clang version 16.0.0
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files\LLVM\bin
clang: error: unsupported option '-fPIC' for target 'x86_64-pc-windows-msvc'
```
Now it's clear.

However the original error from `build2` was quite obscure and felt unrelated, I made other changes before figuring out that it was that issue and would it have been clearer (aka would make me lose less time searching) if the error reported in clang was also displayed as the reason. (also, the "msvc" in that error report was a bit misleading/weird)

Suggestions of improvements (one or the other):

1. When the compiler info gathering fails, look for an error report, if it exists display it too.
2. When the compiler info gathering fails, dump the compiler command's output too.

Of course this is not a massive or urgent issue. I do suspect that if other possible errors can happen in the compiler info extractions step and are not displayed to users properly, a lot of confusion might happen. It's not clear to me if that issue is possible with other toolchains, or if any steps I made was actually supposed to output something different (a bug I'm not aware of).

```
$ b --version
build2 0.16.0-a.0.456e2179dcb1
libbutl 0.16.0-a.0.bb5b2ed9b340
host x86_64-microsoft-win32-msvc14.3
```

**Comment by boris-kolpackov** [comment]
> I see that -fPIC was added to the commands used to retrieve the compiler information. I dont know if it's a bug or not(?) but I assumed it was on purpose for some configuration-specific handling reason.

Yes, the failing command is part of the attempt to detect the C/C++ standard library used and that can be affected by the options. For example:

```
$ b config.cxx=clang++ config.cxx.coptions=-stdlib=libc++
```

Note that the recommended "modern" way to do this is to specify such options (e.g., `-stdlib`, `-m64`, etc) as part of the compiler mode options:

$ b config.cxx="clang++ -stdlib=libc++"

But we still pass `*.coptions` for backwards compatibility with the old way, before we had the compiler mode options. Maybe it's time to drop that, though.

> Suggestions of improvements (one or the other):
>
> 1.  When the compiler info gathering fails, look for an error report, if it exists display it too.
> 2. When the compiler info gathering fails, dump the compiler command's output too.

Yes, the reason it's not done currently is because the information that we are trying to extract is often dumped to `stderr` and is therefore merged with the diagnostics. Your case is actually a good illustration of that. If we were to cache and dump the entire output, it will be:

```
clang version 16.0.0
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files\LLVM\bin
clang: error: unsupported option '-fPIC' for target 'x86_64-pc-windows-msvc'
```

But I agree we need to improve this somehow. The easiest will probably be to print the failing command line which the user can easily re-run. We already do this in a few places.

**Comment by Klaim** [comment]
> But I agree we need to improve this somehow. The easiest will probably be to print the failing command line which the user can easily re-run. We already do this in a few places.

I noted that and I agree it's helpful. :+1:
